internal/fs/buffer/inmemory_write_buffer_test.go

```diff
@@ -41,24 +56,371 @@ func (t *MemoryBufferTest) SetUp(ti *TestInfo) {
+
+	// Write to buffer
+	err := t.mb.WriteAt(data1, 0)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data2, MiB)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data3, 2*MiB)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data4, 3*MiB)
+	AssertEq(nil, err)
+
+	AssertEq(t.mb.fileSize, 4*MiB)
+	AssertTrue(bytes.Equal(data4, t.mb.current.buffer[0:MiB]))
+	AssertTrue(bytes.Equal(data3, t.mb.flushed.buffer[0:MiB]))
+	t.validateInMemoryBuffer(t.mb.current, bufferSizeInMB*MiB, 0, 3*MiB, 4*MiB)
+	t.validateInMemoryBuffer(t.mb.flushed, bufferSizeInMB*MiB, 0, 2*MiB, 3*MiB)
+}
+
+func (t *MemoryBufferTest) TestMultipleRandomWritesToInMemoryBuffer() {
+	t.createAndValidateInMemoryBuffer(bufferSizeInMB)
+
+	// Write to buffer
+	err := t.mb.WriteAt([]byte(smallContent1), 0)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt([]byte(smallContent2), 2)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt([]byte(smallContent3), 7)
+	AssertEq(nil, err)
+
+	AssertEq(t.mb.fileSize, 12)
+	AssertTrue(bytes.Equal([]byte("TaBurriPizza"), t.mb.current.buffer[0:t.mb.fileSize]))
+	t.validateInMemoryBuffer(t.mb.current, bufferSizeInMB*MiB, 0, 0, MiB)
+	t.validateInMemoryBuffer(t.mb.flushed, 0, 0, 0, 0)
+}
+
+func (t *MemoryBufferTest) TestWriteJustBeforeChunkSizeOffset() {
+	t.createAndValidateInMemoryBuffer(bufferSizeInMB)
+
+	// Write to buffer
+	err := t.mb.WriteAt([]byte(smallContent1), MiB-1)
+	AssertEq(nil, err)
+
+	AssertEq(t.mb.fileSize, MiB+3)
+	AssertTrue(bytes.Equal([]byte("T"), t.mb.flushed.buffer[MiB-1:MiB]))
+	AssertTrue(bytes.Equal([]byte("aco"), t.mb.current.buffer[0:3]))
+	t.validateInMemoryBuffer(t.mb.current, bufferSizeInMB*MiB, 0, MiB, 2*MiB)
+	t.validateInMemoryBuffer(t.mb.flushed, bufferSizeInMB*MiB, 0, 0, MiB)
+}
+
+func (t *MemoryBufferTest) TestWriteAtChunkSizeOffset() {
+	t.createAndValidateInMemoryBuffer(bufferSizeInMB)
+
+	// Write to buffer
+	err := t.mb.WriteAt([]byte(smallContent1), MiB)
+	AssertEq(nil, err)
+
+	AssertEq(t.mb.fileSize, MiB+smallContent1Size)
+	AssertTrue(bytes.Equal([]byte(smallContent1), t.mb.current.buffer[0:smallContent1Size]))
+	t.validateInMemoryBuffer(t.mb.current, bufferSizeInMB*MiB, 0, MiB, 2*MiB)
+	t.validateInMemoryBuffer(t.mb.flushed, bufferSizeInMB*MiB, 0, 0, MiB)
+}
+
+func (t *MemoryBufferTest) TestWriteJustAfterChunkSizeOffset() {
+	t.TestCreateEmptyInMemoryBufferWithValidBufferSize()
+
+	// Write to buffer
+	err := t.mb.WriteAt([]byte(smallContent1), MiB+1)
+
+	AssertNe(nil, err)
+	AssertEq(NonSequentialWriteError, err.Error())
+	AssertEq(t.mb.fileSize, 0)
+	t.validateInMemoryBuffer(t.mb.current, MiB, 0, 0, MiB)
+	t.validateInMemoryBuffer(t.mb.flushed, 0, 0, 0, 0)
+}
+
+func (t *MemoryBufferTest) TestWriteRollOverFromCurrentToFlushedBuffer() {
+	t.createAndValidateInMemoryBuffer(bufferSizeInMB)
+	data1 := t.generateRandomData(MiB)
+	data2 := t.generateRandomData(2 * KiB)
+
+	// Write to buffer
+	err := t.mb.WriteAt(data1, 0)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data2, MiB)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data2, 2*MiB-KiB)
```

Comment:
Can we use a data3 instead of data2 here to see that it actually overwrites the previous write (`t.mb.WriteAt(data2, MiB)`) when in overlapping range ?

Reply:
The offsets are entirely different. Data 2 is being written at MiB to MiB + 2KiB and MiB - KiB to MiB + KiB. I don't see a need for this change.

Reply:
Ok, do we have any tests targeting overwriting ? If not, we should.

Reply:
Updated this test to also validate on empty contents. Also caught an error in the clear syntax usage. Thanks 💯